### PR TITLE
Panoc

### PR DESCRIPTION
**Pull request by AjSat**
PANOC now seems to work giving results very similar to the matlab implementation.

### CODE REVIEW

File: source/panoc.c [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+#include "panoc.h"
+
+size_t panoc_algo(panoc_parameters *params, real_t sigma_k, real_t **ux, real_t *x_out, size_t max_iter) {
+    size_t i, j;
+    sigma_k = 0.01; /*Why is this getting overwritten in this function?*/
```
Comment (alphaville) [inline source/panoc.c]: what is this line for?

File: source/panoc.c [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            }
+            box_constraint(params->box, u_new, u_new);
+            copy_vec(u_new, ux[0], params->box->n);
+            fbe_new = fbe_value(*(res[0]), res[1], rk_new, params->box->n, params->box->gamma);
+
+            tau_k = tau_k * 0.5;
+            j++;
+            /*printf("sigma = %f\n", sigma_k);*/
+            // printf("At j = %d, fbe_new = %f < fbe_old - sigma*norm = %f, tau_k = %f\n", j, fbe_new, fbe_old - sigma_k*norm, tau_k);
+        } while (fbe_new > fbe_old - sigma_k * norm && j < 50);
```
Comment (alphaville) [inline source/panoc.c]: `j < 50`: this shouldn't be hard-coded.

File: main.c [hunk 3]
```diff
@@ -227,11 +229,59 @@ int main(int argc, char** argv) {
     print_vec(x_box, box->n);
     printf("proximal(x) = \n");
     print_vec(x_outbox, box->n);
+    
+    /*Test the panoc routine*/
+    panoc_parameters* params = malloc(sizeof(real_t)*NU*NHOR*20); /*Initializing all the
```
Comment (alphaville) [inline main.c]: Wrong allocation: `params` is not of type `real_t` and it doesn't have size equal to `sizeof(real_t)*NU*NHOR*20`.

You can simply do: 

```{.c}
panoc_parameters params;
```

Moreover, `params` is practically unused.

File: main.c [hunk 4]
```diff
@@ -227,11 +229,59 @@ int main(int argc, char** argv) {
...
+                                                        * parameters required to
+                                                     * run PANOC*/
+    
+    params->box = malloc(sizeof(params->box)*NU*NHOR*20); /*Initializing box parameters*/
```
Comment (alphaville) [inline main.c]: Wrong allocation.

File: include/panoc.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ * lbfgs buffer that is required to run the algorithm.
+ * @param sigma - parameters used in line search to check if a step indeed 
+ * causes descent in FBE.
+ * @param xu - an array of arrays, containing the current position x and
+ * u - the initial point from where optimization is iteratively performed.
+ * @param x_out - pointer to an array that stores the value of optimal point
+ * @param max_iter - maximum number of iterations in the line search algorithm
+ * @return 
+ */
+size_t panoc_algo(panoc_parameters *params, real_t sigma, real_t **xu, real_t *x_out, size_t max_iter);
```
Comment (alphaville) [inline include/panoc.h]: A couple of remarks on this function template:

1. Why aren't `sigma` and `max_iter` part of `panoc_parameters`?
2. No information is returned regarding the algorithms such as the number of iterations it took to converge, the status (whether it has converged), the residual and so on.

File: source/panoc.c [hunk 6]
```diff
@@ -0,0 +1,170 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+#include "panoc.h"
+
+size_t panoc_algo(panoc_parameters *params, real_t sigma_k, real_t **ux, real_t *x_out, size_t max_iter) {
```
Comment (alphaville) [inline source/panoc.c]: This function is supposed to return `size_t`, but it doesn't return anything.

File: source/panoc.c [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+    size_t lbfgs_updates = 0;
+    for (i = 0; i < 500; i++) {
+
+        /*printf("Iteration %d:\n", i);*/
+        /*printf("u when computing cost for first time in an iteration = \n");
+        print_vec(ux[0], 4);*/
+        cost((const real_t **) ux, res, iw, w, 0);
+        /*printf("l(u) %f:\n", *(res[0]));*/
+        copy_vec(ux[0], u_old, params->box->n);
```
Comment (alphaville) [inline source/panoc.c]: Use `memcpy` instead of `copy_vec`.

File: source/panoc.c [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        /*printf("norm square of gradient = %f\n",norm_sq(l_grad_old,params->box->n));*/
+        /*printf("then computed gradient = \n");
+        print_vec(res[1], params->box->n);*/
+        old_cost = *(res[0]);
+
+        j = 1;
+        do { /*Implementing backtracking of gamma*/
+
+            if (j > 1) {
+                params->box->gamma = 0.8 * params->box->gamma;
```
Comment (alphaville) [inline source/panoc.c]: `0.8` should not be hard-coded

File: source/panoc.c (hunk 6 above)
Comment (AjSat) [inline source/panoc.c]: I'm returning 0 now.

File: source/panoc.c (hunk 1 above)
Comment (AjSat) [inline source/panoc.c]: The value of sigma in panoc() was some garbage value and not the one that I had sent as a parameter. So, I had reinitialized it while debugging.

File: source/panoc.c [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    real_t *l_grad_new = malloc(sizeof (real_t) * params->box->n);
+    real_t *s = malloc(sizeof (real_t) * params->box->n);
+    real_t *y = malloc(sizeof (real_t) * params->box->n);
+    real_t *rk = malloc(sizeof (real_t) * params->box->n);
+    real_t *rk_new = malloc(sizeof (real_t) * params->box->n);
+    real_t *dk = malloc(sizeof (real_t) * params->box->n);
+    real_t eps = 0.01, l_diff; /*Parameters used for backtracking of gamma*/
+
+    size_t lbfgs_updates = 0;
+    for (i = 0; i < 500; i++) {
```
Comment (AjSat) [inline source/panoc.c]: rectified.

File: source/panoc.c (hunk 2 above)
Comment (AjSat) [inline source/panoc.c]: But, I need to limit the back-tracking on tau somehow or it is getting stuck in a permanent loop if a sufficient decrease in FBE is not observed.

File: main.c (hunk 3 above)
Comment (AjSat) [inline main.c]: That is actually causing the code to crash.

File: source/panoc.c (hunk 2 above)
Comment (panpat) [inline source/panoc.c]: I would suggest a much smaller value for j, e.g., j=5-10. If line search fails then we should use a forward-backward step as back-up (it corresponds to stepsize equal to zero).

File: main.c (hunk 3 above)
Comment (alphaville) [inline main.c]: @AjSat Have a look at [this](https://stackoverflow.com/questions/3774193/constructor-for-structs-in-c).

Comment (AjSat) [inline main.c]: rectified

File: main.c [hunk 10]
```diff
@@ -227,11 +229,59 @@ int main(int argc, char** argv) {
...
+    
+    params->n_hor = NHOR;
+    params->nx = NX;
+    params->nu = NU;
+    params->box->n = NU*NHOR;
+    
+    
+    
+    /*printf("paramsboxn = %d\n", params->box->n);*/
+    params->lbfgs = malloc(sizeof(params->lbfgs)*NU*NHOR*20); /*Initializing LBFGS buffer*/
```
Comment (AjSat) [inline main.c]: done

File: main.c [hunk 11]
```diff
@@ -227,11 +229,59 @@ int main(int argc, char** argv) {
...
+    /*printf("simga in main function = %f\n", sigma);*/
+   /* arg[0][0] = 0.989412732642372; arg[0][1] = 0.821599223180349; arg[0][2] = 5.79809303238556; arg[0][3] = 3.03018953316039; /* 0.989412732642372
+0.821599223180349
+5.79809303238556
+3.03018953316039
+
+
+
+*/
+    panoc_algo(params, sigma, arg, u_out, 1000);
```
Comment (AjSat) [inline main.c]: The cost function is specified in the parameters. I have not implemented that part yet.

File: include/panoc.h (hunk 5 above)
Comment (AjSat) [inline include/panoc.h]: Yes, I need to modify that.

File: main.c [hunk 12]
```diff
@@ -210,28 +194,62 @@ int main(int argc, char** argv) {
...
-    real_t x_box[] = {5, -50}, *x_outbox = malloc(sizeof(real_t)*box->n);
-    box_constraint(box, x_box, x_outbox);
-    printf("The result of proximal operator:\n");
-    printf("x = \n");
-    print_vec(x_box, box->n);
-    printf("proximal(x) = \n");
-    print_vec(x_outbox, box->n);
     return (EXIT_SUCCESS);
-    
+    free(X0);
```
Comment (alphaville) [inline main.c]: These lines are after `return`, so they are never reached.

File: main.c [hunk 13]
```diff
@@ -27,8 +27,10 @@
  */
 int main(int argc, char** argv) {
```
Comment (alphaville) [inline main.c]: A few comments about this method:
- The method is too lengthy, so debugging is very difficult
- The `main` function should not be used for testing (though convenient, there shouldn't be a `main` at all)
- Dynamic memory allocations such as `res[0] = malloc(sizeof (real_t));` should be deallocated using `free(res[0])` -- and `free(res[1])` respectively.

**Pull request closed without merging**